Switch the unbounded dropwizard metrics queues to bounded queues

joshelser opened the pull request:
If the metrics2 subsystem isn't up, we want to make sure we don't
hold an excessive number of metrics in memory.

joshelser committed:
Switch the unbounded dropwizard metrics queues to bounded queues

If the metrics2 subsystem isn't up, we want to make sure we don't
hold an excessive number of metrics in memory.

Closes #3

joshelser commented:
The hope here is that this should not affect current consumers of the reporter.

There are five types of metrics which we cache: gauge, timer, histogram, meter and counter. Each of these "caches" are a queue. Previously the queues were unbounded, but they were switch to be bounded queues now. By default, each queue is bounded to 1000 elements; the total number of metrics is thusly limited to 5000 elements.

When new metrics elements fail to be added to the queue (because the queue is full), the oldest elements are dropped from the queue in order to make room for all new elements. A message is logged at `DEBUG` when this happens.

joshelser committed:
Forgot to add a debug message before we start dropping old metrics from the queue.

sidseth commented on src/main/java/com/github/joshelser/dropwizard/metrics/hadoop/HadoopMetrics2Reporter.java:
```diff
@@ -368,24 +393,57 @@ private void addSnapshot(MetricsRecordBuilder builder, String name, String desc,
+    for (Entry<String,T> entry : metricsToAdd) {
+      // Assume that we have space (normal condition)
+      if (!queue.offer(entry)) {
+        LOG.debug("Failed to add metrics element. Removing {} elements from queue", entriesLeftToAdd);
```
Some of the elements may already have been added. I don't think that is being factored into the removal of elements from the existing queue.

joshelser replied:
Calls to `report()` are already synchronized on `this`. I'm wondering if it would simplify things to also synchronize the `getMetrics()` call in the  same manner. This would prevent any confusion.

It's definitely not ideal WRT performance, but it would at least be correct as a first go. Thoughts on that @sidseth ?

sidseth replied:
Wasn't referring to the synchronization; I think your latest fix is required though :)

It was more about 
When metricsToAdd is larger than maxAllowed - we're guaranteed to hit a queue.offer failure. At that point - the entire queue is effectively cleared (it clears the initial entriesLeftToAdd elements again). 1) a new count of entriesLeftToAdd can be used. 2) Start iterating n elements into the source entries if > entriesLeftToAdd.

Similar cases exist when metricsLeftToAdd is less than the allowed, but the queue is close to it's limit. This can end up deleting more elements than required.

Your offline suggestion about just clearing the queues if there's any overflow may be better. Again - I still don't fully understand the impact since I'm not sure what is reported each time - and time windows being maintained.

joshelser replied:
> When metricsToAdd is larger than maxAllowed - we're guaranteed to hit a queue.offer failure.

Yup, you're right there. Another edge case I could handle better. I started to think about these, but then also wondered how likely it will be that we have that many metrics. I'm not 100% sure either :)

> Your offline suggestion about just clearing the queues if there's any overflow may be better. Again - I still don't fully understand the impact since I'm not sure what is reported each time - and time windows being maintained.

I think what we see in `report()` is essentially the last "period" of time provided by the application (Hive) when the reporter is started. Looks like this is controlled by "hive.service.metrics.hadoop2.frequency" in hive-site (defaulting to 30s). Let me double check this too.

sidseth commented:
Thanks for the quick patch Josh. Have left one comment on the code change.
I don't really understand the impact of the change on the actual metrics. I was of the opinion that the metrics could be aggregated at an interval, or when the bound is hit, instead of dropping them. Looking through the code - don't think that's simple.

joshelser committed:
Also synchronize(this) during the metrics2 collection phase

ScheduledReporter is doing a synchronized(this) before calling
report() on our Reporter implementation. This will make sure there
aren't any unintended side-effects if they are both called at the beginning.

joshelser committed:
Some more clarification on the concurrency side of things

joshelser commented:
How do you feel about the latest additions, @sidseth? Long-term I need to come up with some sort of test-harness for more "real-life" confidence than just the unit tests. I think this will be fine.

joshelser committed:
Address sidseth's suggestion about metrics to add which exceed the maximum limit of the queue

Add in a standalone utility which sets up a FileSink and generates some metrics
to observe the impact of the changes.

joshelser committed:
Move debug logs to trace (not really "debug" appropriate) and fix some javadoc warnings.

joshelser committed:
Some tests for the new builder methods
